File: solutions/115-distinct-subsequences/distinct-subsequences.go
func numDistinct(s string, t string) int {
    M := len(s)
    N := len(t)
    dp := make([]int, N)
    prev := 1
    for i := M - 1; i >= 0; i-- {
        prev = 1
        for j := N - 1; j >= 0; j-- {
            old_dpj := dp[j]
            if s[i] == t[j] {
                dp[j] += prev
            }
            prev = old_dpj
        }
    }
    return dp[0]
}

File: solutions/115-distinct-subsequences/distinct-subsequences.ts
function numDistinct(s: string, t: string): number {
    let M = s.length;
    let N = t.length;
    let dp: number[] = new Array(N).fill(0);
    let prev = 1;
    for (let i = M - 1; i >= 0; i--) {
        prev = 1;
        for (let j = N - 1; j >= 0; j--) {
            let old_dpj = dp[j];
            if (s.charAt(i) == t.charAt(j)) {
                dp[j] += prev;
            }
            prev = old_dpj;
        }
    }
    return dp[0];
}